File: python/pia/resources/sec.py
"""Official PIA SDK - SEC EDGAR Filings Resource."""

from __future__ import annotations

import urllib.parse
from typing import Any, Dict, Optional

from ..errors import ValidationError
from ..transport import AsyncTransport, SyncTransport


class SecResource:
    """Synchronous SEC EDGAR Corporate Filings resource."""

    def __init__(self, transport: SyncTransport) -> None:
        self._transport = transport

    def get_filings(
        self,
        symbol: Optional[str] = None,
        form_type: Optional[str] = None,
        limit: Optional[int] = None,
        since: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Retrieves corporate SEC filings (10-K, 10-Q, 8-K, Form 4 insider transactions)."""
        params: Dict[str, Any] = {}
        if symbol:
            params["symbol"] = symbol.strip().upper()
        if form_type:
            params["form_type"] = form_type
        if limit is not None:
            params["limit"] = limit
        if since:
            params["since"] = since
        return self._transport.request("/api/v1/sec/filings", method="GET", params=params)

    def get_company(self, symbol: str) -> Dict[str, Any]:
        """Retrieves SEC profile and historical filings for a company symbol."""
        if not symbol or not symbol.strip():
            raise ValidationError("Symbol must be a non-empty string.", param_name="symbol")
        clean = symbol.strip().upper()
        encoded = urllib.parse.quote(clean, safe="")
        return self._transport.request(f"/api/v1/sec/companies/{encoded}", method="GET")


class AsyncSecResource:
    """Asynchronous SEC EDGAR Corporate Filings resource."""

    def __init__(self, transport: AsyncTransport) -> None:
        self._transport = transport

    async def get_filings(
        self,
        symbol: Optional[str] = None,
        form_type: Optional[str] = None,
        limit: Optional[int] = None,
        since: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Asynchronously retrieves corporate SEC filings."""
        params: Dict[str, Any] = {}
        if symbol:
            params["symbol"] = symbol.strip().upper()
        if form_type:
            params["form_type"] = form_type
        if limit is not None:
            params["limit"] = limit
        if since:
            params["since"] = since
        return await self._transport.request("/api/v1/sec/filings", method="GET", params=params)

    async def get_company(self, symbol: str) -> Dict[str, Any]:
        """Asynchronously retrieves SEC profile for a company symbol."""
        if not symbol or not symbol.strip():
            raise ValidationError("Symbol must be a non-empty string.", param_name="symbol")
        clean = symbol.strip().upper()
        encoded = urllib.parse.quote(clean, safe="")
        return await self._transport.request(f"/api/v1/sec/companies/{encoded}", method="GET")

File: typescript/src/resources/sec.ts
/**
 * Official PIA SDK - SEC EDGAR Filings Resource
 */

import { ValidationError } from "../errors";
import type { HttpTransport } from "../http/transport";
import type { RequestOptions } from "../types";

export interface SecFilingItem {
  id: string;
  symbol: string;
  form_type: "10-K" | "10-Q" | "8-K" | "4" | "13F" | string;
  company_name: string;
  cik: string;
  filing_date: string;
  report_url: string;
  description?: string;
}

export interface SecFilingsResponse {
  total: number;
  items: SecFilingItem[];
  has_more?: boolean;
}

export interface GetSecFilingsOptions extends RequestOptions {
  symbol?: string;
  form_type?: string;
  limit?: number;
  since?: string;
}

export class SecResource {
  constructor(private readonly transport: HttpTransport) {}

  /**
   * Retrieves latest corporate SEC EDGAR filings (10-K, 10-Q, 8-K, Form 4 insider trades).
   */
  public async getFilings(options?: GetSecFilingsOptions): Promise<SecFilingsResponse> {
    const params = new URLSearchParams();
    if (options?.symbol) params.set("symbol", options.symbol.trim().toUpperCase());
    if (options?.form_type) params.set("form_type", options.form_type);
    if (options?.limit) params.set("limit", String(options.limit));
    if (options?.since) params.set("since", options.since);

    const query = params.toString() ? `?${params.toString()}` : "";
    const raw = await this.transport.request<any>(
      `/api/v1/sec/filings${query}`,
      "GET",
      undefined,
      options
    );
    const items = Array.isArray(raw) ? raw : raw?.items || raw?.filings || raw?.data || [];
    return { total: raw?.total ?? items.length, items: Array.isArray(items) ? items : [], has_more: raw?.has_more };
  }

  /**
   * Retrieves SEC filing history for a specific public company symbol.
   */
  public async getCompany(symbol: string, options?: RequestOptions): Promise<any> {
    if (!symbol || typeof symbol !== "string" || symbol.trim() === "") {
      throw new ValidationError("Symbol must be a non-empty string.", "symbol");
    }
    const cleanSymbol = symbol.trim().toUpperCase();
    return this.transport.request<any>(
      `/api/v1/sec/companies/${encodeURIComponent(cleanSymbol)}`,
      "GET",
      undefined,
      options
    );
  }
}
